packages/teams-js/src/internal/validDomains.ts

```diff
@@ -0,0 +1,104 @@
+import { GlobalVars } from '../internal/globalVars';
+import { validDomainsCdnEndpoint, validOriginsFallback } from './constants';
+import { getLogger } from './telemetry';
+import { isValidHttpsURL } from './utils';
+
+let validOrigins: string[] = [];
+export async function prefetchDomainsFromCDN(): Promise<void> {
+  if (fetch) {
+    await fetch(validDomainsCdnEndpoint);
+  }
+}
+
+async function retrieveDomainsFromCDNAndStore(): Promise<string[]> {
+  if (validOrigins.length !== 0) {
+    return validOrigins;
+  }
+  if (fetch) {
+    return fetch(validDomainsCdnEndpoint)
+      .then((response) => {
+        if (!response.ok) {
+          validOrigins = validOriginsFallback;
+          return validOriginsFallback;
+        }
+        return response.json().then((validDomains) => {
+          validOrigins = validDomains.validOrigins;
+          return validOrigins;
+        });
+      })
+      .catch(() => {
+        validOrigins = validOriginsFallback;
+        return validOrigins;
```

Comment:
This is notable but maybe not interesting. In this failure case you do:
```
        validOrigins = validOriginsFallback;
        return validOrigins;
```
but in the failure case above you do
```
          validOrigins = validOriginsFallback;
          return validOriginsFallback;
```

They seem equivalent but it does make me wonder if it would be possible to unify the failure cases so you aren't doing the same thing twice. E.g., maybe if `!response.ok` you `throw Error("Bad response")` which would then be caught by the same catch block and unify the failure handling?